### DISCUSSION TITLE
Get the list of indirect relationships

### DISCUSSION THREAD

**Discussion by vhpoet** [openfga/community#49] [Q&A] [answered]
I have the below DSL

```
type customer
  relations
    define customer_admin as self
type user
  relations
    define parent as self
type document
  relations
    define admin as customer_admin from parent
    define parent as self
```

and these tuples

```json
{
  "user": "customer:acme",
  "relation": "parent",
  "object": "document:a"
},
{
  "user": "customer:acme",
  "relation": "parent",
  "object": "document:b"
},
{
  "user": "customer:acme",
  "relation": "parent",
  "object": "user:john"
},
{
  "user": "user:john",
  "relation": "customer_admin",
  "object": "customer:acme"
}
```
- `customer:acme` has a `parent` relationship with `user:john`, `document:a` and `document:b`.
- `user:john` has a `customer_admin` relationship with `customer:acme`.
- `user:john` has an indirect `admin` relationship with `document:a` and `document:b`.

Given the situation above, what's the best way to get the list of the documents `user:john` has a direct and indirect `admin` relationship with?

I might be missing something, but I couldn't find a nice way of doing this. What's the general thinking around this? Do I have to replicate some of the relationships in my own database? Are there any resources / best practices I can look at?

Thank you!

**Comment by rhamzeh** [comment, marked as answer]
Hey @vhpoet - thanks for asking!

We list a bit of our thought process on this in the [Search with Permissions](https://openfga.dev/docs/interacting/search-with-permissions) document.

TLDR - It depends on:
- Number of objects that your database can return from a search query
- Number of objects of a certain type the user could have access to
- Percentage of objects in a type the user could have access to

For some of those cases, we have an experimental endpoint called [ListObjects](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects), which given a type, a relation and a user, returns a list of object ids the user can access.

Note that the current approach is experimental and not ready for production use, and comes with a few [caveats](https://openfga.dev/docs/interacting/relationship-queries#caveats-and-when-not-to-use-it-3) including subpar latency at scale. Though we are working on optimizing it and having it production ready as soon as we can.

**Comment by vhpoet** [comment]
Thanks for the quick response! I read that page last week and forgot about it 🤦‍♂️.
